### ISSUE TITLE
Custom maps - 1st implementation

### ISSUE THREAD

**Issue by janagombitova** [akvo/akvo-product-design#143]
### Overview

The custom maps will enable data-driven styling of markers and custom icons on the Akvo FLOW maps. This will expand the basic map functionality to the current FLOW maps and will give the users a chance to understand their collected data better in geographic context. We will implement this feature in separate steps, starting with the most basic functionality and supporting the simplest data structure, followed by tackling more complex items (as monitoring and repeated question groups for example). The users will be able to save their created maps and publish them on their website or share with other partners. The overall feature is explained here: https://github.com/akvo/akvo-product-design/issues/133

The initial implementation will enable to base a map only on option questions per selected registration form. The option question cannot be located in a repeated question group and have multiple select enabled. We will also add a set of permissions covering this new feature.
### Links to documents

Mock ups
https://github.com/akvo/akvo-flow/files/64147/Custom.maps.-.1st.Implementation.pdf

Functional design document
https://github.com/akvo/akvo-product-design/blob/master/FLOW/Features/133-Custom-maps/1st%20implementation%20-%20Functional%20design%20document.md

Implementation feedback
https://docs.google.com/document/d/1mTKqOshhDVT47KCci2OnJBUdiho54eSXKwi6zRRSHzM/edit

Tests 
https://docs.google.com/document/d/1uPZPJqPweTK_Rnlx4s0ULrv294e_owSEHjJd5VS_fqc/edit

**Comment by janagombitova** [comment]
#### Planning:
* [ ] implement custom custom maps management (new subtab, create, save, edit, view and delete a custom map) 
* [ ] enable colour coding of data points in new map based on options 
* [ ] validate that only option questions are provided which are not in repeated question groups and do not have multiple select enabled 
* [ ] validate that with a monitored form selected, there is no question selector dropdown 
* [ ] implement colour picker 
* [ ] implement legend
* [ ] improve UI 
* [ ] user permissions to cover 1st implementation 
* [ ] add time range as a filter

**Comment by janagombitova** [comment]
## Current state 

We have the initial prototype of this functionality was built. This feature is currently **on hold** due to the below mentioned reasons. 

Actions that still need to happen are: code review and feature review. If we are to implement this functionality in Flow, it needs to solve a problem our users have and support them in how they work with maps when capturing data. Thus we have conducted a set of user testing interviews to: 
1) evaluate the implementation of the feature
2) understand how users use maps currently, what benefits do they bring them

Secondly, releasing this feature is currently **on hold**. This is due to the above mentioned actions that still need to happen, but also due to a technical hurdle that needs to be overcome. The Custom maps are currently tied with the integration of Carto maps. After initial testing of the implementation, we found that there are problems with our data pipeline. These issues are handled under the Reflow project. Only once this is solved, we can continue with further integrating Carto maps to Flow's online workspaces (instances).  

### Testing custom maps

Here are the learning points that we gathered over the last months working on the Flow Custom maps prototype and testing it with users. 

#### What have we done so far? 
- Gathered user stories from the partner team and users on (1) how users use Flow data and visualise it on maps and (2) how do they use the current maps in Flow - https://docs.google.com/document/d/1A-XrNF9N5QvBxFuFEB9CsV9LOxA4A1pvPzPbDZqRXYI/edit#heading=h.ga98yqqzhiyv
- Created a prototype of a new feature "custom maps” based on these stories and contractual obligations using CartoDB as a framework and example, build by Charles, without code review (further work is currently paused on this feature) - https://drive.google.com/drive/folders/0B53AXQ5Y0VCMZmd0VUtaNWM0YUU
- Tested the prototype with 6 current Flow users, lead by Anne, summarised the learning points from the user testing - https://peak.akvo.org/results-of-user-testing-flow-maps-24-august-2016/

#### What did we learn? 

The learning points shared below are a summary of insights gather from the user testing of the Custom maps prototype. Reflecting on the insights that came out of the user testing we can say that the way maps can be used in Flow covers a different user need than how they can be used in Lumen. 

**Users:**  
We identified two types of users: Field staff who use the maps to track the progress of data collection projects (usually based in the location of the project and in direct contact with enumerators) and Programme officers or M&E officers (usually based in HQs) who use maps to understand collected data and share the visuals with others. 

**Flow learning points:**

**Current use:** 
- users do not use Flow maps at this moment  
   - the cluster set up does not give them any useful information and they cannot recognise which point belongs to which project
   - the location of the clusters may give a “incorrect” view of location of data points, 
   - hard to see details about a point, readability of the answers in the side panel is not sufficient, you need to zoom and click on the point to learn more   
   - the maps lack interactive options, can only use for viewing the overall scale of data collection, but nothing more
   - many did not know that by zooming into the map the separate points start showing
   - the public map is not used at all either and users do not know its purpose 
- some users use the maps functionality to monitor the progress of data collection exercises
   - users zoom into the location of where the data collection should be taking place on a daily basis a visually check if points are there or not and how many
   - they randomly open the point and look at the answers, if they are submitted correctly and if not, then search for the point in raw data, find the enumerator, cross check his other submissions and contact him 

**Need:** 
*(Flow as a data collection tool and not a data visualisation tool)* 
- be able to see data points based on a survey
- user who does not have view data rights for a survey, must not be able to see the data on the map
- user needs to be able to identify the submission easily
- be able to see points based on submission time - to be able to track if data is coming in the proper location on time < monitor progress of data collection activities 
- be able to see who submitted the points, or even see points based on submitter < for monitoring data collection activities

**Comment by janagombitova** [comment]
Based on the following reasons, we will not continue with this effort: 
- user research of the prototype even though showed this is a nice addition to have, our users use the maps for different purposes, mostly to track progress of data collection, and for these situations we will implement functionality that can ease their work. Having custom visualisations is something used for reporting and decisions making
- part of Akvo's mission and strategy is not to build what already exists, by building a data visualisation tool for maps we will build what is already provided by many open source and not open source tools also used by our partners   
- Akvo is currently developing a data visualisation tool called Akvo Lumen, which main purpose is visualising data. Lumen's direct connection with Flow and free price tier for all Flow users will allow our users to use Lumen's visualisation features (also maps). By working on integrating the two tools, we will not be duplicating efforts around features that are supported in one or the other tool. 
- Akvo Flow is a data capturing tool not a data visualisation tool and we aim to develop functionality that supports our users in capturing data
